src/plugins/intel_npu/src/plugin/npuw/compiled_model.hpp

```diff
@@ -178,10 +179,41 @@ class CompiledModel : public ov::npuw::ICompiledModel {
         // FIXME: This is a 1:1 copy of the ov::npuw::Subgraph structure
         // w.r.t. function calls
         std::size_t param_base = 0;
+
+        // Need to wrap closure, since finalize_weights_bank() will
+        // asynchronously evaluate weights and put them in closure.
+        // Other functions of CompiledModel as well as InferRequest and
+        // other entities need to wait for the closure to be populated first
+        // (meaning to wait for async weights processing to end).
+        class SafeClosureWrapper {
+        public:
+            std::vector<ov::Tensor>& unsafe_get_closure() {
+                return m_closure;
+            }
+            std::vector<ov::Tensor>& get_closure() {
+                if (m_evaluated) {
+                    return m_closure;
+                }
+                if (m_evaluation.valid()) {
+                    m_evaluation.wait();
+                    m_evaluated = true;
+                }
+                return m_closure;
+            }
+            void set_future(std::shared_future<void>& evaluation) {
+                m_evaluation = evaluation;
+            }
+
+        private:
+            std::vector<ov::Tensor> m_closure;
+            std::shared_future<void> m_evaluation;
+            bool m_evaluated = false;
+        };
+        mutable SafeClosureWrapper closure;
```

Comment:
There's no need to make this field mutable, it should maintain the same contracts as the original vector.

Reply:
Done